spirv_msl.cpp

```diff
@@ -6489,6 +6549,31 @@ void CompilerMSL::fix_up_shader_inputs_outputs()
 					          to_expression(builtin_subgroup_invocation_id_id), " - 32, 0)), uint2(0));");
 				});
 				break;
+			case BuiltInViewIndex:
+				if (!msl_options.multiview)
+				{
+					// According to the Vulkan spec, when not running under a multiview
+					// render pass, ViewIndex is 0.
+					entry_func.fixup_hooks_in.push_back([=]() {
+						statement("const ", builtin_type_decl(bi_type), " ", to_expression(var_id), " = 0;");
+					});
+				}
+				else if (get_execution_model() == ExecutionModelVertex)
+				{
+					// Metal provides no special support for multiview, so we smuggle
+					// the view index in the instance index.
+					entry_func.fixup_hooks_in.push_back([=]() {
+						statement(builtin_type_decl(bi_type), " ", to_expression(var_id), " = ",
+						          msl_options.multiview_base_view, " + ", to_expression(builtin_instance_idx_id), " % ",
```

Comment:
So, I guess the idea here is that you want to detect all contiguous ranges in the view mask, and render all those in one go?

As layered rendering is used here, how will this work with loadOp = CLEAR? Can you restrict the render pass to just base layer + count? Is the [[render_target_index]] shifted to accomodate this?

Reply:
There's also the issue of indirect draws ...

Reply:
> So, I guess the idea here is that you want to detect all contiguous ranges in the view mask, and render all those in one go?

Pretty much.

> As layered rendering is used here, how will this work with loadOp = CLEAR? Can you restrict the render pass to just base layer + count? Is the [[render_target_index]] shifted to accomodate this?

I can set the `renderTargetArrayLength` property in the `MTLRenderPassDescriptor` to set the number of views to be rendered, and the `slice` property of the `MTLRenderPassAttachmentDescriptor` to set the base view. This is the same as regular layered drawing.

> Does this mean we need different shaders to deal with different parts of the render pass where the base layer needs to change?

I'm considering reverting the change to use hard-coded constants.

> There's also the issue of indirect draws ...

We already have to munge the indirect buffer for tessellated draws. I can probably do that for multiview, too.

Reply:
Ok, I guess if you're munging the indirect buffers, this will work.